File: Python/mpi.py
from mpi4py import MPI
import time
from cannon import *
from math import sqrt
from matrix_basic import read_matrix_from_file

@timeit
def main():
    comm = MPI.COMM_WORLD

    comm.Barrier()
    wt = MPI.Wtime()
    rank = comm.Get_rank()
    size = comm.Get_size()
    processorName = MPI.Get_processor_name()

    BLOCKS = int(sqrt(size))

    for k in range(BLOCKS+1):
        i = int(rank / BLOCKS)
        j = int(rank % BLOCKS)
        if k == 0:
            if rank == 0:
                A = read_matrix_from_file('input/matrixA')

                #B = read_matrix_from_file('input/matrixB')

                SUB_ELEM = int(len(A)/BLOCKS)

                ma = divide_into_blocks(A, BLOCKS)
                #mb = divide_into_blocks(B, BLOCKS)
                ma = initial_shift_left(ma)
                #mb = initial_shift_up(mb)

                for n in range(BLOCKS):
                    for m in range(BLOCKS):
                        if n == 0 and m == 0:
                            continue
                        comm.send(ma[n][m], dest=(n * BLOCKS + m), tag=1)
                        #comm.send(mb[n][m], dest=(n * BLOCKS + m), tag=2)

                dataA = ma[0][0]
                #dataB = mb[0][0]
                dataB = comm.recv(source=1, tag=2)
                C = [[0 for i in range(SUB_ELEM)] for j in range(SUB_ELEM)]
            elif rank == 1:
                B = read_matrix_from_file('input/matrixB')
                SUB_ELEM = int(len(B) / BLOCKS)
                mb = divide_into_blocks(B, BLOCKS)
                mb = initial_shift_up(mb)
                for n in range(BLOCKS):
                    for m in range(BLOCKS):
                        if n == 0 and m == 1:
                            continue
                        comm.send(mb[n][m], dest=(n * BLOCKS + m), tag=2)
                dataB = mb[0][1]
                dataA = comm.recv(source=0, tag=1)
                C = [[0 for i in range(SUB_ELEM)] for j in range(SUB_ELEM)]
            else:
                dataA = comm.recv(source=0, tag=1)
                dataB = comm.recv(source=1, tag=2)
                SUB_ELEM = int(len(dataA))
                C = [[0 for i in range(SUB_ELEM)] for j in range(SUB_ELEM)]

        else:
            # izracunaj
            C = multiple_matrixes(dataA, dataB, C, 0, 0)

            comm.send(dataA, dest=(i * BLOCKS + (j + BLOCKS - 1) % BLOCKS), tag=1)
            comm.send(dataB, dest=(((i+BLOCKS-1) % BLOCKS)*BLOCKS + j), tag=2)

            dataA = comm.recv(source=(i * BLOCKS + (j + 1) % BLOCKS), tag=1)
            dataB = comm.recv(source=(((i + 1) % BLOCKS)*BLOCKS + j), tag=2)

    if rank == 0:
        blocks = [[] for i in range(size)]
        blocks[0] = C
        for i in range(1, size):
            blocks[i] = comm.recv(source=i)

        res = [[0 for i in range(len(A))] for j in range(len(A[0]))]
        res = blocks_to_matrix(blocks, res)
        write_matrix_to_file("output/parallelOutput", res)
    else:
        comm.send(C, dest=0)


    comm.Barrier()

    if rank == 0:
        wt = MPI.Wtime() - wt
        print("Vreme: ", wt)


main()

File: Go/mpi.go
package main

import (
	"fmt"
	"matrix_basic"
	"cannon"
	"sync"
	"time"
	"math"
	"os"
	"strconv"
)

func ccx(wg *sync.WaitGroup, c chan [][]int, x [][]int, id int){
	defer wg.Done()
	c <- x
	//c <- y
	//fmt.Println(x, id)
}

func ccc(c chan [][]int, n int){
	var x, y [][]int
	select {
		case x = <-c:
			select {
			case y = <-c:
				fmt.Println("YES")
			}
		default:
			fmt.Println("NO")
	}
	
	
	fmt.Println(n,x, y)
}

func process(wg *sync.WaitGroup, process_id int, chanA, chanB []chan [][]int, C *[][]int) {
	
	defer wg.Done()

	BLOCKS := int(math.Sqrt(float64(len(chanA))))
	i := int(process_id / BLOCKS)
	j := int(process_id % BLOCKS)
	var dataA, dataB [][]int
	var ok bool
	
	for r:= 0; r <BLOCKS; r++{
		select {
			// preuzimamo blok matrice A
			case dataA, ok = <-(chanA)[process_id]:
				if ok {
					//fmt.Printf("Value %d was read.%d\n", dataA, process_id)

				} else {
					fmt.Println("Channel closed! Something went wrong")
				}
		}

		select {
			// preuzimamo blok matrice B
			case dataB, ok = <-(chanB)[process_id]:
				if ok {
					//fmt.Printf("Value %d was read.%d\n", dataB, process_id)

				} else {
					fmt.Println("Channel closed! Something went wrong")
			}
		}
		*C, _ = matrix_basic.Multiple_matrixes(dataA, dataB, *C, i*len(dataA),j*len(dataA))
		//fmt.Println(process_id,C)

		select {
			// prosledjujemo sledecem procesu blok matrice A
			case chanA[i * BLOCKS + (j + BLOCKS - 1)%BLOCKS] <- dataA:
		}


		select {
			// prosledjujemo sledecem procesu blok matrice B
			case chanB[((i + BLOCKS - 1) % BLOCKS)*BLOCKS + j] <- dataB:
	}
	}
}

 
func main(){

	start := time.Now()

	temp := os.Args[1]

	PROCESS, err := strconv.Atoi(temp)
	var wg, wg2 sync.WaitGroup

	A, err :=  matrix_basic.Read_matrix_from_file("input/matrixA")
	if (err != nil){
		fmt.Println("Something went wrong")
	}

	B, err := matrix_basic.Read_matrix_from_file("input/matrixB")
	if (err != nil){
		fmt.Println("Something went wrong")
	}

	BLOCKS := int(math.Sqrt(float64(PROCESS)))

	//submatrix_elements := int(len(A) / BLOCKS)
    ma := cannon.Divide_into_blocks(A, BLOCKS)
    mb := cannon.Divide_into_blocks(B, BLOCKS)

	C := matrix_basic.Initialize_matrix(len(A), len(A[0]))
	
	cannon.Initial_shift_left(ma)
	cannon.Initial_shift_up(mb)


	var chansA []chan [][]int
	var chansB []chan [][]int
	for i :=0 ; i< PROCESS;i++ {
		//chansA[i] = make(chan [][]int, 1)
		chansA = append(chansA, make(chan [][]int,1))
		chansB = append(chansB, make(chan [][]int,1))
		//chansB[i] = make(chan [][]int, 1)
		if i < BLOCKS*BLOCKS{
			wg2.Add(1)
			go ccx(&wg2, chansA[i], ma[i/BLOCKS][i%BLOCKS], i)
			wg2.Add(1)
			go ccx(&wg2, chansB[i], mb[i/BLOCKS][i%BLOCKS], i)
		}
	}
	wg2.Wait()
	for r:=0; r<BLOCKS*BLOCKS;r++{
		wg.Add(1)
		go process(&wg, r, chansA, chansB, &C)
	}

	wg.Wait()
	matrix_basic.Write_matrix_to_file("output/parallelOutput", &C)
	fmt.Println("Vreme: ", time.Since(start))
}